ocrd_models/ocrd_page_user_methods.py

```diff
@@ -83,6 +83,26 @@ def show(self):
 #
 # Replace the following method specifications with your own.
 
+#
+# List all *Regions on the PAGE
+#
+get_AllRegions = MethodSpec(name='get_AllRegions',
+    source=r'''
+    def get_AllRegions(self, regions=None, reading_order=False):
+        """"
+        Get all the *Region element or only those provided by ``regions``.
+        Returned in random order unless ``reading_order`` is set (NOT CURRENTLY IMPLEMENTED)
+        """
+        if reading_order:
+            raise NotImplementedError("Ordering of regions by Reading Order not currently Implemented :(")
+        if not regions:
+            regions = ['Advert', 'Chart', 'Chem', 'Custom', 'Graphic', 'Image', 'LineDrawing', 'Map', 'Maths', 'Music', 'Noise', 'Table', 'Text']
+        ret = []
+        for region in regions:
+            ret += getattr(self, 'get_{}Region'.format(region))()
+        return ret
```

Comment:
Have not tried it out, but how about this:

```suggestion
        if reading_order:
            reading_order = reading_order.get_OrderedGroup() or reading_order.get_UnorderedGroup()
        if reading_order:
            def get_recursive_reading_order(rogroup):
                if isinstance(rogroup, (OrderedGroupType, OrderedGroupIndexedType)):
                    elements = rogroup.get_AllIndexed()
                if isinstance(rogroup, (UnorderedGroupType, UnorderedGroupIndexedType)):
                    elements = (rogroup.get_RegionRef() + rogroup.get_OrderedGroup() + rogroup.get_UnorderedGroup())
                regionrefs = list()
                for elem in elements:
                    regionrefs.append(elem.get_regionRef())
                    if not isinstance(elem, (RegionRefType, RegionRefIndexedType)):
                        regionrefs.extend(get_recursive_reading_order(elem))
                return regionrefs
            reading_order = get_recursive_reading_order(reading_order)
        if reading_order:
            ret = dict([(region.id, region) for region in ret])
            return [ret[region_id] for region_id in reading_order if region_id in ret]
        else:
            return ret
```

Since the ReadingOrder tree is an injective mapping to the region tree, but we are only interested in the top-level regions here, we have to first get the full sequence and then "filter".